### PR TITLE
[core] Use bcrypt instead of MariaDB PASSWORD() + automatic migration

### PR DESCRIPTION
**Pull request by zach2good**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Stops using Maria's `PASSWORD()` for hashing and storage in the db. Instead uses bcrypt.
- Offers an automatic migration path for users. New accounts will start using bcrypt right away, users logging in or changing their PW will be migrated to the new system seamlessly.
- Update CI in the one place where we make a fake account to login

:star: another PR while I'm packed into a train under the channel tunnel

### CODE REVIEW

**Comment by zach2good** [comment]
Testing:

1:
- Create character before changes, login
- Apply changes, rebuild
- Log in, see hash change to bcrypt hash in the db

2:
- With changes applied
- Create new character
- See new character's PW hash in db is bcrypt style

3:
- Roll back changes, rebuild, make new character, see they have old-style hash
- Change their password with the old system, see hash changes
- Apply changes, rebuild
- Do password reset flow
- Log in
- See hash is bcrypt style

**Comment by zach2good** [comment]
^ For each of these, get your PW wrong and you should be left in your original state, not in some floating state where you need admin intervention

File: src/login/auth_session.cpp [hunk 1]
```diff
@@ -309,13 +365,68 @@ void auth_session::read_func()
...
+                if (ret != SQL_ERROR && sql->NumRows() != 0 && sql->NextRow() == SQL_SUCCESS)
+                {
+                    return sql->GetStringData(0);
+                }
+                return "";
+            }();
+            // clang-format on
+
+            // OLD PASSWORD HASH MIGRATION
+            // TODO: Remove me after 2023-12-31
```
Comment (TeoTwawki) [inline src/login/auth_session.cpp]: did you mean 2024?

Comment (zach2good) [reply src/login/auth_session.cpp]: Yessir

File: src/login/auth_session.cpp [hunk 2]
```diff
@@ -138,9 +140,65 @@ void auth_session::read_func()
...
+                if (ret != SQL_ERROR && sql->NumRows() != 0 && sql->NextRow() == SQL_SUCCESS)
+                {
+                    return sql->GetStringData(0);
+                }
+                return "";
+            }();
+            // clang-format on
+
+            // OLD PASSWORD HASH MIGRATION
+            if (std::string(__DATE__) == "Dec 31 2024")
```
Comment (WinterSolstice8) [inline src/login/auth_session.cpp]: Wouldn't this only trip if your build date was specifically december 31st 2024 and not any days after?

Comment (zach2good) [reply src/login/auth_session.cpp]: Quite right, I've added a constexpr check and a static_assert to blow things up if this section is still here when 2025 starts. Tested by rolling the target year to 2023 and it stops compiling producing the help message

**Review by Xaver-DaRed — APPROVED** [review]

**Pull request merged**